File: common-ui-list/file-system/src/main/java/com/storyteller_f/file_system/instance/local/DocumentLocalFileInstance.java
package com.storyteller_f.file_system.instance.local;

import android.content.Context;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.documentfile.provider.DocumentFile;

import com.storyteller_f.file_system.FileSystemUriSaver;
import com.storyteller_f.file_system.instance.BaseContextFileInstance;
import com.storyteller_f.file_system.instance.Create;
import com.storyteller_f.file_system.instance.FileCreatePolicy;
import com.storyteller_f.file_system.instance.NotCreate;
import com.storyteller_f.file_system.model.DirectoryItemModel;
import com.storyteller_f.file_system.model.FileItemModel;
import com.storyteller_f.file_system.util.FileInstanceUtility;
import com.storyteller_f.file_system.util.FileUtility;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

import kotlin.Pair;

public class DocumentLocalFileInstance extends BaseContextFileInstance {
    private static final String TAG = "DocumentLocalFileInstan";
    public static final String EXTERNAL_STORAGE_DOCUMENTS = "com.android.externalstorage.documents";

    protected DocumentFile current;

    /**
     * 用来标识对象所在区域，可能是外部，也可能是内部
     * 比如/storage/XXXX-XXXX
     */
    String prefix;
    /**
     * 一般是authority
     */
    private String preferenceKey;

    private String pathRelativeRoot;

    public DocumentLocalFileInstance(Context context, Uri uri, String preferenceKey, String prefix) {
        super(context, uri);
        this.prefix = prefix;
        assertInit();
        this.preferenceKey = preferenceKey;
        current = getDocumentFile();
    }

    private DocumentLocalFileInstance(Context context, Uri uri, String prefix) {
        super(context, uri);
        this.prefix = prefix;
        assertInit();
    }

    private void assertInit() {
        assert getPath().startsWith(prefix);
        if (Objects.equals(getPath(), prefix)) {
            pathRelativeRoot = "/";
        } else {
            pathRelativeRoot = getPath().substring(prefix.length());
        }
        assert pathRelativeRoot.startsWith("/");

    }

    /**
     * 初始化DocumentFile，初始化失败一般不影响获取目录,但是不可以对当前对象进行操作
     */
    public DocumentFile getDocumentFile() {
        return getDocumentFile(NotCreate.INSTANCE);
    }

    /**
     * 获取指定目录的document file
     *
     * @return 返回目标文件
     */
    private DocumentFile getDocumentFile(FileCreatePolicy policy) {
        //此uri 是当前前缀下的根目录uri。fileInstance 的uri 是fileSystem 使用的uri。
        Uri rootUri = FileSystemUriSaver.getInstance().savedUri(context, preferenceKey);
        if (rootUri == null) {
            Log.e(TAG, "getDocumentFile: rootUri is null");
            return null;
        }
        DocumentFile rootFile = DocumentFile.fromTreeUri(context, rootUri);
        if (rootFile == null) {
            Log.e(TAG, "getDocumentFile: fromTreeUri is null");
            return null;
        }
        if (!rootFile.canRead()) {
            Log.e(TAG, "getDocumentFile: 权限过期, 不可读写 " + getPath() + " prefix: " + prefix);
            return null;
        }

        if (pathRelativeRoot.equals("/")) return rootFile;
        String[] nameItemPath = pathRelativeRoot.substring(1).split("/");
        String[] paths;
        boolean endElementIsFileName = policy instanceof Create && ((Create) policy).isFile();
        if (endElementIsFileName) {
            paths = Arrays.copyOf(nameItemPath, nameItemPath.length - 1);
        } else paths = nameItemPath;

        DocumentFile temp = rootFile;
        for (String name : paths) {
            if (needStop()) break;
            DocumentFile foundFile = temp.findFile(name);
            if (foundFile == null) {
                if (policy instanceof NotCreate) {
                    Log.e(TAG, "getDocumentFile: 文件找不到" + getPath() + " prefix: " + prefix);
                    return null;
                }
                DocumentFile created = temp.createDirectory(name);
                if (created == null) {
                    Log.e(TAG, "getDocumentFile: 文件创建失败" + getPath() + " prefix: " + prefix);
                    return null;
                } else temp = created;
            } else temp = foundFile;
        }

        //find file
        if (endElementIsFileName) {
            String fileName = nameItemPath[nameItemPath.length - 1];
            DocumentFile file = temp.findFile(fileName);
            if (file == null) {
                return temp.createFile("*/*", fileName);
            } else return file;
        }
        return temp;
    }

    @Override
    public long getDirectorySize() {
        return getDocumentFileSize(current);
    }

    private long getDocumentFileSize(DocumentFile documentFile) {
        long size = 0;
        DocumentFile[] documentFiles = documentFile.listFiles();
        for (DocumentFile documentFi : documentFiles) {
            if (needStop()) break;
            if (documentFile.isFile()) {
                size += documentFi.length();
            } else {
                size += getDocumentFileSize(documentFi);
            }
        }
        return size;
    }

    @Override
    public String getParent() {
        DocumentFile parentFile = current.getParentFile();
        if (parentFile == null) {
            return null;
        }
        return parentFile.getUri().getPath();
    }

    public boolean createDirectory() {
        if (current != null) return true;
        DocumentFile created = getDocumentFile(new Create(false));
        if (created != null) {
            current = created;
            return true;
        }
        return false;
    }

    public boolean createFile() {
        if (current != null) return true;
        DocumentFile created = getDocumentFile(new Create(true));
        if (created != null) {
            current = created;
            return true;
        }
        return false;
    }

    @Override
    public BaseContextFileInstance toChild(@NonNull String name, FileCreatePolicy policy) throws Exception {
        if (!exists()) {
            Log.e(TAG, "toChild: 未经过初始化或者文件不存在：" + getPath());
            return null;
        }
        if (isFile()) {
            throw new Exception("当前是一个文件，无法向下操作");
        } else {
            Uri build = uri.buildUpon().path(new File(getPath(), name).getAbsolutePath()).build();
            DocumentLocalFileInstance instance = new DocumentLocalFileInstance(context, build, prefix);
            instance.preferenceKey = preferenceKey;
            instance.current = getChild(name, policy);
            return instance;
        }
    }

    /**
     * @param name 名称
     * @return 如果查找不到，而且不用创建，返回null
     * @throws Exception 会出现无法预计的结果时，不允许再次继续
     */
    public DocumentFile getChild(String name, FileCreatePolicy policy) throws Exception {
        DocumentFile file = current.findFile(name);
        if (file != null) {
            return file;
        } else if (!(policy instanceof Create)) {
            return null;
        } else if (((Create) policy).isFile()) {
            DocumentFile createdFile = current.createFile("*/*", name);
            if (createdFile != null) {
                return createdFile;
            } else {
                throw new Exception("创建文件失败");
            }
        } else {
            DocumentFile createdDirectory = current.createDirectory(name);
            if (createdDirectory != null) {
                return createdDirectory;
            } else {
                throw new Exception("创建文件夹失败");
            }
        }
    }

    @Override
    public void changeToChild(@NonNull String name, FileCreatePolicy policy) throws Exception {
        if (isFile()) {
            throw new Exception("当前是一个文件，无法向下操作");
        } else {
            this.current = getChild(name, policy);
        }
    }

    @Override
    public void listInternal(@NonNull List<FileItemModel> files, @NonNull List<DirectoryItemModel> directories) throws Exception {
        if (current == null) {
            current = getDocumentFile();
        }
        DocumentFile c = current;
        if (c == null) throw new Exception("no permission");
        DocumentFile[] documentFiles = c.listFiles();
        for (DocumentFile documentFile : documentFiles) {
            if (needStop()) break;
            String documentFileName = documentFile.getName();
            assert documentFileName != null;
            String detailString = FileUtility.getPermissions(documentFile);
            Pair<File, Uri> t = child(documentFile, documentFileName);
            if (documentFile.isFile()) {
                FileInstanceUtility.addFile(files, t, detailString).setSize(documentFile.length());
            } else {
                FileInstanceUtility.addDirectory(directories, t, detailString);
            }
        }
    }

    @NonNull
    private Pair<File, Uri> child(DocumentFile documentFile, String documentFileName) {
        Pair<File, Uri> child = child(documentFileName);
        return new Pair<>(new File(child.getFirst().getAbsolutePath()) {
            @Override
            public long lastModified() {
                return documentFile.lastModified();
            }

            @Override
            public long length() {
                return documentFile.length();
            }
        }, child.getSecond());
    }

    @Override
    public boolean deleteFileOrEmptyDirectory() {
        return current.delete();
    }

    @Override
    public BaseContextFileInstance toParent() throws Exception {
        File parentFile = new File(getPath()).getParentFile();
        if (parentFile == null) {
            throw new Exception("到头了，无法继续向上寻找");
        }
        DocumentFile currentParentFile = current.getParentFile();
        if (currentParentFile == null) {
            throw new Exception("查找parent DocumentFile失败");
        } else if (!currentParentFile.isFile()) {
            Uri parent = uri.buildUpon().path(parentFile.getAbsolutePath()).build();
            DocumentLocalFileInstance instance = new DocumentLocalFileInstance(context, parent, prefix);
            instance.preferenceKey = preferenceKey;
            instance.current = currentParentFile;
            return instance;
        } else {
            throw new Exception("当前文件已存在，并且类型不同 源文件：" + currentParentFile.isFile());
        }

    }

    @Override
    public void changeToParent() throws Exception {
        File parentFile = new File(getPath()).getParentFile();
        if (parentFile == null) {
            throw new Exception("无法继续向上寻找");
        }
        DocumentFile documentFile = current.getParentFile();
        if (documentFile == null) {
            throw new Exception("查找parent DocumentFile失败");
        } else if (!documentFile.isFile()) {
            current = documentFile;
        } else {
            throw new Exception("当前文件已存在，并且类型不同 源文件：" + documentFile.isFile());
        }

    }

    @Override
    public boolean isFile() {
        if (current == null) {
            Log.e(TAG, "isFile: path:" + getPath());
        }
        return current.isFile();
    }

    @Override
    public boolean exists() {
        if (current == null) {
            return false;
        }
        return current.exists();
    }

    @Override
    public boolean isDirectory() {
        if (current == null) {
            Log.e(TAG, "isDirectory: isDirectory:" + getPath());
        }
        return current.isDirectory();
    }

    @Override
    public void changeTo(@NonNull String path) {
        if (this.getPath().equals(path)) {
            return;
        }
        if (path.startsWith(prefix)) {
            current = getDocumentFile();
        }
    }

    @Override
    public boolean rename(String newName) {
        return current.renameTo(newName);
    }

    @Override
    public boolean isHidden() {
        return false;
    }

    @Override
    public FileInputStream getFileInputStream() throws FileNotFoundException {
        ParcelFileDescriptor r = context.getContentResolver().openFileDescriptor(current.getUri(), "r");
        return new FileInputStream(r.getFileDescriptor());
    }

    @Override
    public FileOutputStream getFileOutputStream() throws FileNotFoundException {
        return new FileOutputStream(context.getContentResolver().openFileDescriptor(current.getUri(), "w").getFileDescriptor());
    }

    @Override
    public long getFileLength() {
        return current.length();
    }

    @Override
    public FileItemModel getFile() {
        return new FileItemModel(getName(), uri, false, current.lastModified(), false, FileUtility.getExtension(getName()));
    }

    @Override
    public DirectoryItemModel getDirectory() {
        return new DirectoryItemModel(getName(), uri, false, current.lastModified(), false);
    }

    public static DocumentLocalFileInstance getEmulated(Context context, Uri uri, String prefix) {
        return new DocumentLocalFileInstance(context, uri, EXTERNAL_STORAGE_DOCUMENTS, prefix);
    }

    /**
     * sd 卡使用特殊的preferenceKey，就是路径
     */
    public static DocumentLocalFileInstance getMounted(Context context, Uri uri, String prefix) {
        //fixme sdCard 与emulated authority 相同，只是rootId 不同
        return new DocumentLocalFileInstance(context, uri, prefix, prefix);
    }
}

File: common-ui-list/file-system/src/main/java/com/storyteller_f/file_system/instance/local/DocumentLocalFileInstance.kt
package com.storyteller_f.file_system.instance.local

import android.annotation.SuppressLint
import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.documentfile.provider.DocumentFile
import com.storyteller_f.file_system.FileSystemUriSaver
import com.storyteller_f.file_system.instance.BaseContextFileInstance
import com.storyteller_f.file_system.instance.FileCreatePolicy
import com.storyteller_f.file_system.instance.FileCreatePolicy.*
import com.storyteller_f.file_system.instance.FileInstance
import com.storyteller_f.file_system.instance.GetDocumentFile
import com.storyteller_f.file_system.model.DirectoryItemModel
import com.storyteller_f.file_system.model.FileItemModel
import com.storyteller_f.file_system.util.FileInstanceUtility
import com.storyteller_f.file_system.util.FileUtility
import kotlinx.coroutines.yield
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream

/**
 * 如果是通过挂载路径 访问/storage/emulated/0，uri 是file:///storage/emulated/0/Downloads，prefix 是/storage/emulated/0，tree 是primary:
 * 如果是通过DocumentProvider 访问前者，uri 是content://authority/primary:/Downloads，prefix 是primary:，tree 是primary:
 * @param prefix 用来标识对象所在区域，便于从path 中截取root对应的真正的路径。
 * 如果是内存卡就是/storage/XXXX-XXXX。
 * 如果是mounted，就是/storage/emulated/0。
 * 如果是DocumentProvider，就是/ 加上treeId，并且treeId可以是/。内存卡和mounted 也可以通过DocumentProvider 直接访问
 * @param preferenceKey 一般是authority，用于获取存储在FileSystemUriSaver 中取出rootUri。
 * 如果是DocumentProvider，tree 也是必须的。
 * @param tree 是DocumentProvider 的treeId，用来区分不同DocumentProvider 多个根的情况
 */
@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
class DocumentLocalFileInstance(
    private val prefix: String,
    private val preferenceKey: String,
    private val tree: String,
    context: Context,
    uri: Uri
) : BaseContextFileInstance(context, uri) {
    private var _instance: DocumentFile? = null

    private suspend fun getInstanceRelinkIfNeed(): DocumentFile? {
        val temp = _instance
        if (temp == null) {
            _instance = (getDocumentFile(NotCreate) as? GetDocumentFile.Success)?.file
        }
        return _instance
    }

    /**
     * 有些document provider 的uri 对应的路径不是path，所以需要从中截取真正的路径
     * 比如mounted 路径，完整路径是/storage/emulated/0/Downloads，
     * prefix 是/storage/emulated/0，
     * path 是/storage/emulated/0/Downloads，
     * pathRelativeRoot 是/Downloads
     */
    private val pathRelativeRoot: String by lazy {
        if (path == prefix) {
            "/"
        } else {
            path.substring(prefix.length)
        }.apply {
            assert(startsWith("/"))
        }
    }

    init {
        assert(path.startsWith(prefix))
    }

    /**
     * 获取指定目录的document file
     *
     * @return 返回目标文件
     */
    private suspend fun getDocumentFile(policy: FileCreatePolicy): GetDocumentFile {
        // 此uri 是当前前缀下的根目录uri。fileInstance 的uri 是fileSystem 使用的uri。
        val rootUri = FileSystemUriSaver.instance.savedUri(context, preferenceKey, tree)
            ?: return GetDocumentFile.Failed(Exception("rootUri is null"))
        val rootFile = DocumentFile.fromTreeUri(context, rootUri)
            ?: return GetDocumentFile.Failed(Exception("fromTreeUri is null"))
        if (!rootFile.canRead()) {
            return GetDocumentFile.Failed(Exception("权限过期, 不可读写"))
        }
        if (pathRelativeRoot == "/") return GetDocumentFile.Success(rootFile)
        val nameItemPath = pathRelativeRoot.substring(
            1
        ).split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val createdLastFileIfNeed = policy is Create && policy.isFile
        val paths = if (createdLastFileIfNeed) {
            nameItemPath.copyOfRange(0, nameItemPath.size - 1)
        } else {
            nameItemPath
        }
        return loopDocumentFile(
            rootFile,
            paths,
            policy,
            if (createdLastFileIfNeed) nameItemPath.last() else null
        )
    }

    private suspend fun loopDocumentFile(
        rootFile: DocumentFile,
        paths: Array<String>,
        policy: FileCreatePolicy,
        fileName: String?
    ): GetDocumentFile {
        var temp: DocumentFile = rootFile
        for (name in paths) {
            yield()
            val foundFile = temp.findFile(name)
            temp = when {
                foundFile != null -> foundFile
                policy is NotCreate -> {
                    return GetDocumentFile.Failed(Exception("文件找不到$path prefix: $prefix"))
                }

                else -> temp.createDirectory(name)
                    ?: return GetDocumentFile.Failed(Exception("文件创建失败$path prefix: $prefix"))
            }
        }

        return if (fileName == null) {
            GetDocumentFile.Success(temp)
        } else {
            val file = temp.findFile(fileName)
            if (file != null) {
                GetDocumentFile.Success(file)
            } else {
                val created: DocumentFile? = temp.createFile("*/*", fileName)
                if (created != null) {
                    GetDocumentFile.Success(created)
                } else {
                    GetDocumentFile.Failed(Exception("创建失败"))
                }
            }
        }
    }

    override suspend fun getDirectorySize(): Long = getDocumentFileSize(getInstanceRelinkIfNeed())

    private suspend fun getDocumentFileSize(documentFile: DocumentFile?): Long {
        var size: Long = 0
        val documentFiles = documentFile!!.listFiles()
        for (documentFi in documentFiles) {
            yield()
            size += if (documentFile.isFile) {
                documentFi.length()
            } else {
                getDocumentFileSize(documentFi)
            }
        }
        return size
    }

    override suspend fun createDirectory(): Boolean {
        if (getInstanceRelinkIfNeed() != null) return true
        val created = tryCreate(false)
        if (created != null) {
            _instance = created
            return true
        }
        return false
    }

    override suspend fun createFile(): Boolean {
        if (getInstanceRelinkIfNeed() != null) return true
        val created = tryCreate(true)
        if (created != null) {
            _instance = created
            return true
        }
        return false
    }

    private suspend fun tryCreate(isFile: Boolean) =
        (getDocumentFile(Create(isFile)) as? GetDocumentFile.Success)?.file

    @Throws(Exception::class)
    override suspend fun toChild(name: String, policy: FileCreatePolicy): FileInstance? {
        if (!exists()) {
            Log.e(TAG, "toChild: 未经过初始化或者文件不存在：$path")
            return null
        }
        if (isFile()) {
            throw Exception("当前是一个文件，无法向下操作")
        }

        val build = uri.buildUpon().path(File(path, name).absolutePath).build()
        val instance = DocumentLocalFileInstance(prefix, preferenceKey, tree, context, build)
        instance._instance = getChild(name, policy)
        return instance
    }

    /**
     * @param name 名称
     * @return 如果查找不到，而且不用创建，返回null
     * @throws Exception 会出现无法预计的结果时，不允许再次继续
     */
    @Throws(Exception::class)
    suspend fun getChild(name: String?, policy: FileCreatePolicy?): DocumentFile? {
        val instanceRelinkIfNeed = getInstanceRelinkIfNeed()
        val file = instanceRelinkIfNeed!!.findFile(name!!)
        return file
            ?: if (policy !is Create) {
                null
            } else if (policy.isFile) {
                val createdFile = instanceRelinkIfNeed.createFile("*/*", name)
                createdFile ?: throw Exception("创建文件失败")
            } else {
                val createdDirectory = instanceRelinkIfNeed.createDirectory(name)
                createdDirectory ?: throw Exception("创建文件夹失败")
            }
    }

    @Throws(Exception::class)
    public override suspend fun listInternal(
        fileItems: MutableList<FileItemModel>,
        directoryItems: MutableList<DirectoryItemModel>
    ) {
        val c = getInstanceRelinkIfNeed() ?: throw Exception("no permission")
        val documentFiles = c.listFiles()
        for (documentFile in documentFiles) {
            yield()
            val documentFileName = documentFile.name!!
            val detailString = FileUtility.getPermissions(documentFile)
            val t = child(documentFile, documentFileName)
            if (documentFile.isFile) {
                FileInstanceUtility.addFile(fileItems, t, detailString)!!.size =
                    documentFile.length()
            } else {
                FileInstanceUtility.addDirectory(directoryItems, t, detailString)
            }
        }
    }

    private fun child(documentFile: DocumentFile, documentFileName: String): Pair<File, Uri?> {
        val child = child(documentFileName)
        return Pair<File, Uri?>(
            object : File(child.first.absolutePath) {
                override fun lastModified(): Long {
                    return documentFile.lastModified()
                }

                override fun length(): Long {
                    return documentFile.length()
                }
            },
            child.second
        )
    }

    override suspend fun deleteFileOrEmptyDirectory(): Boolean {
        return getInstanceRelinkIfNeed()!!.delete()
    }

    @Throws(Exception::class)
    override suspend fun toParent(): BaseContextFileInstance {
        val parentFile = File(path).parentFile ?: throw Exception("到头了，无法继续向上寻找")
        val currentParentFile = getInstanceRelinkIfNeed()!!.parentFile
        checkNotNull(currentParentFile) {
            "查找parent DocumentFile失败"
        }
        check(currentParentFile.isFile) {
            "当前文件已存在，并且类型不同 源文件：" + currentParentFile.isFile
        }

        val parent = uri.buildUpon().path(parentFile.absolutePath).build()
        val instance = DocumentLocalFileInstance(prefix, prefix, tree, context, parent)
        instance._instance = currentParentFile
        return instance
    }

    override suspend fun isFile(): Boolean {
        val instanceRelinkIfNeed = getInstanceRelinkIfNeed()
        if (instanceRelinkIfNeed == null) {
            Log.e(TAG, "isFile: path:$path")
        }
        return instanceRelinkIfNeed!!.isFile
    }

    override suspend fun exists(): Boolean {
        return getInstanceRelinkIfNeed()?.exists() ?: false
    }

    override suspend fun isDirectory(): Boolean {
        val instanceRelinkIfNeed = getInstanceRelinkIfNeed()
        if (instanceRelinkIfNeed == null) {
            Log.e(TAG, "isDirectory: isDirectory:$path")
        }
        return instanceRelinkIfNeed!!.isDirectory
    }

    override suspend fun rename(newName: String): Boolean {
        return getInstanceRelinkIfNeed()!!.renameTo(newName)
    }

    override suspend fun isHidden(): Boolean = false

    @SuppressLint("Recycle")
    @Throws(FileNotFoundException::class)
    override suspend fun getFileInputStream(): FileInputStream {
        val r = context.contentResolver.openFileDescriptor(getInstanceRelinkIfNeed()!!.uri, "r")
        return FileInputStream(r!!.fileDescriptor)
    }

    @SuppressLint("Recycle")
    @Throws(FileNotFoundException::class)
    override suspend fun getFileOutputStream(): FileOutputStream = FileOutputStream(
        context.contentResolver.openFileDescriptor(
            getInstanceRelinkIfNeed()!!.uri,
            "w"
        )!!.fileDescriptor
    )

    override suspend fun getFileLength(): Long = getInstanceRelinkIfNeed()!!.length()

    override suspend fun getFile(): FileItemModel {
        return FileItemModel(
            name,
            uri,
            false,
            getInstanceRelinkIfNeed()!!.lastModified(),
            false,
            FileUtility.getExtension(name)
        )
    }

    override suspend fun getDirectory(): DirectoryItemModel {
        return DirectoryItemModel(
            name,
            uri,
            false,
            getInstanceRelinkIfNeed()!!.lastModified(),
            false
        )
    }

    companion object {
        private const val TAG = "DocumentLocalFileInstan"
        const val EXTERNAL_STORAGE_DOCUMENTS = "com.android.externalstorage.documents"
        const val EXTERNAL_STORAGE_DOCUMENTS_TREE = "primary:"

        /**
         * 不是通过DocumentProvider 直接访问的。
         */
        fun getEmulated(context: Context, uri: Uri, prefix: String): DocumentLocalFileInstance {
            assert(uri.scheme == ContentResolver.SCHEME_FILE)
            return DocumentLocalFileInstance(
                prefix,
                EXTERNAL_STORAGE_DOCUMENTS,
                EXTERNAL_STORAGE_DOCUMENTS_TREE,
                context,
                uri
            )
        }

        /**
         * 不是通过DocumentProvider 直接访问的。
         * sd 卡使用特殊的preferenceKey，正常来说就是路径
         */
        fun getMounted(context: Context, uri: Uri, prefix: String): DocumentLocalFileInstance {
            assert(uri.scheme == ContentResolver.SCHEME_FILE)
            val tree = getMountedTree(prefix)
            return DocumentLocalFileInstance(prefix, EXTERNAL_STORAGE_DOCUMENTS, tree, context, uri)
        }

        fun getMountedTree(prefix: String) = prefix.substring(prefix.lastIndexOf("/") + 1)

        fun uriFromAuthority(authority: String, tree: String): Uri {
            return Uri.Builder().scheme(ContentResolver.SCHEME_CONTENT).authority(authority)
                .path("/$tree")
                .build()
        }
    }
}
